ssm/variational.py

```diff
@@ -283,75 +326,118 @@ def sample_discrete_states(self):
                 for prms in self.params]
 
     def sample_continuous_states(self):
-        return [block_tridiagonal_sample(prms["J_diag"], prms["J_lower_diag"], prms["h"])
+        return [kalman_info_sample(prms["J_ini"], prms["h_ini"], 0,
+                                   prms["J_dyn_11"], prms["J_dyn_21"], prms["J_dyn_22"],
+                                   prms["h_dyn_1"], prms["h_dyn_2"], 0,
+                                   prms["J_obs"], prms["h_obs"], 0)
                 for prms in self.params]
 
     def sample(self):
         return list(zip(self.sample_discrete_states(), self.sample_continuous_states()))
 
+    def _cache_expectations(self):
+        """Cache continuous and discrete expectations.
+
+        Discrete states are stored as _mean_discrete_states and continuous states
+        as _continuous_expectations. This is so we can re-use the output of hmm_discrete_states()
+        and kalman_info_smoother() if the parameters used to calculate them have not changed.
+        """
+        self._mean_discrete_states = [hmm_expected_states(prms["pi0"], prms["Ps"], prms["log_likes"])
+                for prms in self.params]
+        self._continuous_expectations = [kalman_info_smoother(prms["J_ini"], prms["h_ini"], 0,
+                                     prms["J_dyn_11"], prms["J_dyn_21"], prms["J_dyn_22"],
+                                     prms["h_dyn_1"], prms["h_dyn_2"], 0,
+                                     prms["J_obs"], prms["h_obs"], 0)
+                for prms in self.params]
+        self._prev_params = deepcopy(self._params)
+        
+
     @property
+    @cache_variational_posterior_expectations
     def mean_discrete_states(self):
-        # Now compute the posterior expectations of z under q(z)
-        return [hmm_expected_states(prms["pi0"], prms["Ps"], prms["log_likes"])
-                for prms in self.params]
+        return self._mean_discrete_states
 
     @property
+    @cache_variational_posterior_expectations
     def mean_continuous_states(self):
         # Now compute the posterior expectations of z under q(z)
-        return [block_tridiagonal_mean(prms["J_diag"], prms["J_lower_diag"], prms["h"], lower=True).reshape(np.shape(prms["h"]))
-                for prms in self.params]
+        full_expectations = self.continuous_expectations
+        return [exp[1] for exp in full_expectations]
+
+    @property
+    @cache_variational_posterior_expectations
+    def continuous_expectations(self):
+        return self._continuous_expectations
 
     @property
+    @cache_variational_posterior_expectations
```

Comment:
Is this one necessary?  Looks like the cache will be updated when `mean_discrete_states` and `mean_continuous_states` are updated.

Reply:
indeed! thanks.